### ISSUE TITLE
Custom loss/likelihood formulation with masking

### ISSUE THREAD

**Issue by srn284** [cornellius-gp/gpytorch#1366]
With a deterministic neural network, I conduct multitask learning on some input x on 2 tasks. The basic architecture in code is:
x = featureextractor(x)
out1 = outFNN1(x)
out2 = outFNN2(x)

However, where out1 has a label, out2 does not and vice versa. So for gpytorch multitask learning I have a multitask framework copied and pasted below. But I want to select only some x for the first output and the rest for the second. Thank you

```
num_tasks=2
latentstate=17
inducingPt=16
class MultitaskGPModel(gpytorch.models.ApproximateGP):
    def __init__(self, inducing_points):
        # We have to mark the CholeskyVariationalDistribution as batch
        # so that we learn a variational distribution for each task
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
            inducing_points.size(-2), batch_shape=torch.Size([num_tasks]))
        # We have to wrap the VariationalStrategy in a MultitaskVariationalStrategy
        # so that the output will be a MultitaskMultivariateNormal rather than a batch output
        variational_strategy = gpytorch.variational.MultitaskVariationalStrategy(
            gpytorch.variational.VariationalStrategy(
                self, inducing_points, variational_distribution, learn_inducing_locations=True
            ), num_tasks=num_tasks)
        super().__init__(variational_strategy)
        # The mean and covariance modules should be marked as batch
        # so we learn a different set of hyperparameters
        self.mean_module = gpytorch.means.ConstantMean(batch_shape=torch.Size([num_tasks]))
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([num_tasks])),
            batch_shape=torch.Size([num_tasks])
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        sout = gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
        return sout
inducing_points = torch.rand(num_tasks, inducingPt, latentstate)
model = MultitaskGPModel(inducing_points)
likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=num_tasks)
if torch.cuda.is_available():
    model = model.to("cuda:0")
    likelihood = likelihood.to("cuda:0")







import os
smoke_test = ('CI' in os.environ)
num_epochs = 1 if smoke_test else 20


model.train()
likelihood.train()

# We use SGD here, rather than Adam. Emperically, we find that SGD is better for variational regression
optimizer = torch.optim.Adam([
    {'params': model.parameters()},
    {'params': likelihood.parameters()},
], lr=0.01)

# Training loader
train_loader = torch.utils.data.DataLoader(torch.utils.data.TensorDataset(train_x, train_y), batch_size=10)

# Our loss object. We're using the VariationalELBO, which essentially just computes the ELBO
mll = gpytorch.mlls.VariationalELBO(likelihood, model, num_data=train_y.size(0), combine_terms=False)

# We use more CG iterations here because the preconditioner introduced in the NeurIPS paper seems to be less
# effective for VI.
for i in range(15):
    # Within each iteration, we will go over each minibatch of data
    minibatch_iter = tqdm.tqdm_notebook(train_loader, desc=f"(Epoch {i}) Minibatch")
    for x_batch, y_batch in minibatch_iter:
        optimizer.zero_grad()
#         x_batch = x_batch.reshape(17,7)
#         print(x_batch.shape)                
        x_batch, y_batch = x_batch.to("cuda:0"), y_batch.to("cuda:0")
        output = model(x_batch)
#         print(likelihood(output).mean.shape)
        log_lik, kl_div, log_prior = mll(output, y_batch)
        loss = -(log_lik - kl_div + log_prior)

        minibatch_iter.set_postfix(loss=loss.item())
        loss.backward()
        optimizer.step()

```

Of course this piece of code does multitask learning but likelihood returns loss on 2 tasks for ALL inputs. Even if i split up loss into the loglik and the kl parts, I don't know how to _not_ consider some inputs x' for loss.
Some ways (maybe?) :
1) of course, two different kernels for each task
2) with multitask, some form of masking on likelihood formulation? not sure of course. perhaps with variational sgp, might be hard to do so?
3) any type of multitask with different number of inputs in each task.

Also I am using:
Pytorch 1.4.0
gpytorch 1.1.1

Any way we can try and find a solution in this 1.1.1 version? 

Thank you very much and sorry if it's a repost/duplicate!

**Comment by jacobrgardner** [comment]
You can just mask out log likelihood terms that you don't want to train through before calling backward. You don't have to do anything with the KL or log prior terms, as those are independent of the inputs.

To do this the easiest way, I would just extend VariationalELBO to MaskedVariationalELBO (or the likelihood itself and modify expected_log_likelihood) and modify the line of code here:
https://github.com/cornellius-gp/gpytorch/blob/13edd025de59cc2fef5c18bf2b87c932a1fdb19d/gpytorch/mlls/variational_elbo.py#L61

To basically mask out terms you don't want before calling the final `sum(-1)`. Note that since kwargs make it in to that call, you can define a `mask=None` kwarg so that when you do `mll(output, y_batch, mask=...)` you'll get the mask in that method.

**Comment by srn284** [comment]
Ah thank you for the swift response. Really appreciate it :) 
I upgraded to 1.2.1 by the way and pytorch 1.6.

However, I'm running into a problem when I extend the VariationalELBO class and edit the line above: if target output was shape: [100,4] i.e. 100 data points with 4 task outputs for each point, I am able to mask the first dimension before the sum() call but i wish to mask the second dimension, which is size 4, and I might be mistaken but i don't think changing the line above will help me on this.

Instead I looked at likelihood and changed the expected_log_prob as you were saying above. I have attached a snippet below: 

```
class Gauss2(Likelihood):
    """Base class for Gaussian Likelihoods, supporting general heteroskedastic noise models."""

    def __init__(self, noise_covar: Noise, **kwargs: Any) -> None:

        super().__init__()
        param_transform = kwargs.get("param_transform")
        if param_transform is not None:
            warnings.warn(
                "The 'param_transform' argument is now deprecated. If you want to use a different "
                "transformaton, specify a different 'noise_constraint' instead.",
                DeprecationWarning,
            )

        self.noise_covar = noise_covar

    def _shaped_noise_covar(self, base_shape: torch.Size, *params: Any, **kwargs: Any):
        return self.noise_covar(*params, shape=base_shape, **kwargs)

    def expected_log_prob(self, target: Tensor, input: MultivariateNormal, *params: Any, **kwargs: Any) -> Tensor:
        mean, variance = input.mean, input.variance
        num_event_dim = len(input.event_shape)
        noise = self._shaped_noise_covar(mean.shape, *params, **kwargs).diag()
        # Potentially reshape the noise to deal with the multitask case
        noise = noise.view(*noise.shape[:-1], *input.event_shape)
#         print(noise.shape)
        res = ((target - mean) ** 2 + variance) / noise + noise.log() + math.log(2 * math.pi)
        res = res.mul(-0.5)
        mask = torch.tensor([1,0,1,0]).repeat(100,1)
        res = res * mask
        if num_event_dim > 1:  # Do appropriate summation for multitask Gaussian likelihoods
            res = res.sum(list(range(-1, -num_event_dim, -1)))
        return res

    def forward(self, function_samples: Tensor, *params: Any, **kwargs: Any) -> base_distributions.Normal:
        noise = self._shaped_noise_covar(function_samples.shape, *params, **kwargs).diag()
        return base_distributions.Normal(function_samples, noise.sqrt())

    def log_marginal(
        self, observations: Tensor, function_dist: MultivariateNormal, *params: Any, **kwargs: Any
    ) -> Tensor:
        marginal = self.marginal(function_dist, *params, **kwargs)
        # We're making everything conditionally independent
        indep_dist = base_distributions.Normal(marginal.mean, marginal.variance.clamp_min(1e-8).sqrt())
        res = indep_dist.log_prob(observations)

        # Do appropriate summation for multitask Gaussian likelihoods
        num_event_dim = len(function_dist.event_shape)
        if num_event_dim > 1:
            res = res.sum(list(range(-1, -num_event_dim, -1)))
        return res

    def marginal(self, function_dist: MultivariateNormal, *params: Any, **kwargs: Any) -> MultivariateNormal:
        mean, covar = function_dist.mean, function_dist.lazy_covariance_matrix
        noise_covar = self._shaped_noise_covar(mean.shape, *params, **kwargs)
        full_covar = covar + noise_covar
        return function_dist.__class__(mean, full_covar)
```

And I make a custom Likelihood module with this Gauss2 extension of the Likelihood class. I think I did it ok but again, I might be doing something wrong, so if you could please check it out and let me know if there is anything afoul that would be great. I attached a notebook as well below so you can run the multitask example given.
[multitaskGP_originaltutorial.txt](https://github.com/cornellius-gp/gpytorch/files/5609213/multitaskGP_originaltutorial.txt) (But please change the .txt to .ipynb) 

Thank you again!

**Comment by jacobrgardner** [comment]
Yep, that seems reasonable.

This should work just as is with minibatch training -- just make the mask the appropriate size for the minibatch.

By the way, if you define your method:
```python
def expected_log_prob(self, target: Tensor, input: MultivariateNormal, mask=None, *params: Any, **kwargs: Any) -> Tensor:
```
then you can directly use `mask` without having to index in to `kwargs`

**Comment by srn284** [comment]
Thanks for the brisk reply. Ah yes that's a good point. I should rewrite the method call as above. 

One last q: if I rewrite the MultitaskGP to be able to take inputs with hidden dimensions like below (input is [batch, hidden] and output is [batch, 4]), everything runs perfectly with masked variational elbo with no errors or anything but is my formulation of hidden state inputs bad/not kosher with gpytorch code base? 

hidden dimension is 17 in this case and i reformulate the inducing point as: 

`inducing_points = torch.rand(num_tasks, inducingPt, latentstate)
`


Full code of the extended approximateGP model below:

```
num_tasks=4
latentstate=17
inducingPt=16

class MultitaskGPModel(gpytorch.models.ApproximateGP):
    def __init__(self, inducing_points):
        # We have to mark the CholeskyVariationalDistribution as batch
        # so that we learn a variational distribution for each task
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
            inducing_points.size(-2), batch_shape=torch.Size([num_tasks])
        )

        # We have to wrap the VariationalStrategy in a MultitaskVariationalStrategy
        # so that the output will be a MultitaskMultivariateNormal rather than a batch output
        variational_strategy = gpytorch.variational.IndependentMultitaskVariationalStrategy(
            gpytorch.variational.VariationalStrategy(
                self, inducing_points, variational_distribution, learn_inducing_locations=True
            ),
            num_tasks=4,
        )

        super().__init__(variational_strategy)

        # The mean and covariance modules should be marked as batch
        # so we learn a different set of hyperparameters
        self.mean_module = gpytorch.means.ConstantMean(batch_shape=torch.Size([num_tasks]))
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([num_tasks])),
            batch_shape=torch.Size([num_tasks])
        )

    def forward(self, x):
        # The forward function should be written as if we were dealing with each output
        # dimension in batch
#         print(x.shape)
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        sout = gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
        return sout


# The shape of the inducing points should be (2 x m x 1) - so that we learn different inducing
# points for each output
inducing_points = torch.rand(num_tasks, inducingPt, latentstate)
model = MultitaskGPModel(inducing_points)

```

Again, i appreciate the time @jacobrgardner that you've given for this, and I hope that this last question is not a repost/duplicate. Thank you very much!

**Comment by jacobrgardner** [comment]
@srn284 Looks fine to me.

**Comment by srn284** [comment]
Thank you very much again Jacob for all your help.
